### DISCUSSION TITLE
blog/zh/vector-database/

### DISCUSSION THREAD

**Discussion by giscus** [guangzhengli/blog-comments#11] [Announcements]
# blog/zh/vector-database/

Ladder@也许你最近可能听过这样的新闻，某向量数据库的初创公司刚写好 PPT，就获得了几千万的投资，某公司的开源的向量数据库因其代码的简陋而登上了 Hackernews 等等。在过去几个月时间中， AI 应用的发展如火如荼，带动了 AI 应用技术栈上下游的火爆，而向量数据库就是其中最热门的之一。
笔者最近因为开发 ChatFiles 和 VectorHub 两款开源项目的需要从而对向量数据库（Vector Database）进行了学习，在对主流的向量数据库和搜索算法有了大概的了解后，笔者决定将这些知识整理成一篇文章，希望能够帮助到大家。

https://guangzhengli.com/blog/zh/vector-database/

**Comment by fycycycyc** [comment]
太神了太详细了！！！不是cs专业的也看懂大部分了！！！深入浅出，层层递进，看过的最好的技术博客之一！！！

**Reply by guangzhengli** [reply]
感谢喜欢🩷

**Reply by Junyi-99** [reply]
我特地登录进来给你点赞

**Reply by ahu** [reply]
我也是

**Reply by jiangren1254** [reply]
我也是

**Comment by windmaple** [comment]
向量数据库+ANN并不是什么崭新的领域，做推荐系统的人都不知道用了多少年了，只是知道的人相对少而已，这一波儿被 LLM 带红了

**Reply by guangzhengli** [reply]
感谢提醒，说的非常对。向量数据库应用在传统的人脸识别、图像搜索等领域其实已经有比较长的时间，但个人觉得向量数据库的商业化和对于 LLM 实际场景落地方面还算是一个比较新的领域。

**Reply by baiwfg2** [reply]
确实，我看ANN论文最早可追溯到几十年前了，确实是老东西，没想到现在才为DB人所知

**Comment by foolyoghurt** [comment]
非常好的技术科普文章，embedding 虽然不是新技术，但是以前确实不知道，因为工作无关。现在大家都在讨论，并且借助 OpenAI 的 API 可以较低成本的完成很多有趣的事情，这一波实属是 OpenAI 的技术普惠了

**Comment by zhugaojian** [comment]
讲的太好了

**Comment by meyergin** [comment]
画图用的是什么app呀？

**Reply by guangzhengli** [reply]
VS Code + [Excalidraw](https://excalidraw.com) + [Excalidraw Libraries](https://libraries.excalidraw.com)

**Reply by meyergin** [reply]
谢谢分享

**Comment by lulyon** [comment]
感谢，很完备的梳理。虽然ann相关的技术（各个技术点单独使用，没有把它看作数据库）都用过，但是和传统数据库结合起来，是一个很好的产品化的视角。

**Comment by noneback** [comment]
质量好高！关注了！

**Comment by mpgood** [comment]
非常好的文章，赞！

**Comment by marsbaiyun** [comment]
写的太好了！

**Comment by limboinf** [comment]
我想知道博主用的是哪个向量数据库？目前我用的是chroma，虽然入门很简单，但是其分布式及副本容灾能力不太行，还不是太成熟。

**Reply by AwesomeYuer** [reply]
chroma正准备换掉clickhouse，感觉产品还没定型，milvus(zillia）产品最完善，分布式，高可用群集，qdrant 的github stars最近长得挺快

**Reply by guangzhengli** [reply]
我为了开发方便，暂时使用的是 pgvector。但我暂时推荐 qdrant，目前感觉还挺好的，1GB 集群免费使用，集成起来也还行，Rust 写加上社区也还行。

**Comment by AwesomeYuer** [comment]
pgvector extension 属于暴力检索，基于倒排转置索引 ivf，而且承诺召回率，准确率100%

其他专业向量数据库还支持至少hnsw

**Reply by AwesomeYuer** [reply]
专业向量索引基本都是列存储索引

**Comment by mwnu** [comment]
好文章，比较了多种搜索算法的优缺点，这让我产生了一个问题：
是否存在使用注意力机制的搜索算法？根据上下文和自然语境搜索相似性时给与某些向量的维度更高优先权和计算权重？

**Comment by ycjcl868** [comment]
深入浅出，好文！

**Comment by mushanzi1216** [comment]
将的很详细，最近再看llm相关的文章，对向量数据库的认知还很模糊，这篇文章解答了我的疑惑。
另外想问下有使用过pinecone嘛，为什么不建议使用这个

**Comment by Diomchen** [comment]
写的太棒了！

**Comment by foreverpiano** [comment]
有没有rss提供选项，想订阅文章!!

**Comment by Brandon-lz** [comment]
chroma是可以云端部署的，可以server和client分离，具体看https://docs.trychroma.com/usage-guide

**Comment by Asunfly** [comment]
![](https://storage.guangzhengli.com/images/lsh1.png)
请问这张图是不是不太对呀，理解到这里没理解清楚。
第一位坐标，ABCD分别为：0010说明这是横线
第二位坐标，ABCD分别为：1101说明这是横线
第三位坐标，ABCD分别为：0011说明这是纵线
第四位坐标，ABCD分别为：0010说明这是横线
按照这种应该是三条横线分开才对？希望大佬解答一下，感谢感谢。

**Reply by 17336** [reply]
向量D的哈希值应该是0111

**Reply by guangzhengli** [reply]
感谢两位的疑问和解答，也提醒了我这个图可以进行优化，最后只留下两条直线，现在应该比较好理解。

![result](https://storage.guangzhengli.com/images/lsh2.png)

**Comment by 17336** [comment]
写的很好啊！但是LSH中那张图，向量D的哈希值应该是0111吧

**Comment by ysgncss** [comment]
一起发送给GPT的是什么呢？向量怎么和问题一起喂给GPT？

**Reply by guangzhengli** [reply]
向量只是用来检索，最后从数据库中查出来的还是文本，所以最后发给GPT的是文本。

**Comment by jo32** [comment]
写得太棒了，逻辑清晰，深入浅出。留意到您也在开发 ChatFiles，是基于 next.js 的一个项目，不妨试一下这款向量数据库 🙏：

https://closevector.getmegaportal.com/

https://closevector-docs.getmegaportal.com/blog/release

我是 CloseVector 的作者，这款向量数据库相比起一般向量数据库的区别是它针对中小量条数目场景，将数据库下载到用户本地运行（目前提供了 web 版本）和 CDN 存储服务（当然也可以自行解决 CDN 存储），好处是不耗费服务端资源和伸缩性较好，比较适合 Chat With PDF 等场景。

希望和您能有进一步的交流 😄

**Comment by Jinxuyang** [comment]
点赞

**Comment by nobody1986** [comment]
hnsw刚看到的时候也是感觉想到了跳表。

**Comment by scoMz** [comment]
特地登陆点赞

**Comment by dushulin** [comment]
给作者手动点赞，同时想请教下有哪些可以端侧部署的向量数据库推荐呢？

**Comment by sissilab** [comment]
非常感谢您的分享，关于如何将文本（尤其大文本）转embedding，想请教下如何能在相似匹配和控制token数中间找到平衡：
1. 仅对 title 转 embedding：token数很少，但丢失content信息
2. 集合title+content先利用GPT反向生成问题，然后对这些问题+title生成embedding：生成embedding开销很小，但利用GPT反向生成问题时存在大量损耗
3. 集合title+content先利用GPT进行summarize，然后对summary生成embedding：生成embedding开销较小，但利用GPT在summarize时存在大量损耗
4. 对title+content（截取前n个句子，默认认为越往前的内容越重要）生成embedding：生成embedding开销一般，存在内容信息丢失
5. 对title+content（chunk）生成embedding：生成embedding开销大，依赖chunk方法

不知有无该方面的心得能否分享下，主要是：如何既能保证高相似匹配，又能控制开销？若考虑chunk content，如何chunk？

**Comment by zjackli** [comment]
特地登录进来给你点赞, 世界需要更多像博主一样的老师

**Comment by codexvn** [comment]
太吊了，后端人受益颇多

**Comment by RuixuanYing** [comment]
这篇文章写的非常棒！要是作者能提供PD F版本就更好了，方便做笔记和思考

**Comment by spccold** [comment]
好文

**Comment by zhuxixi** [comment]
周六加班到晚上，回家前竟然能看到这么好的文章，被作者感动到泪目了。

**Comment by chllei** [comment]
非常感谢！写得逻辑清晰，收获非常大。

**Comment by qiaohaoforever** [comment]
感谢公司论坛的兄弟写了一篇好文章，链接了你这么高质量的blog，收藏了～
我也喜欢写blog，向你学习！

**Comment by Olivia-account** [comment]
我也是特意进来给你点赞

**Comment by fonnywong** [comment]
写得太棒了，爱了爱了，原理部分讲得非常明白，感谢后面列出的参考，期待后续的文章，这篇已经收藏了

**Comment by CodeLaurie** [comment]
有个问题，如果通过向量数据库构建搜索系统，进行多特征匹配的话，如何对query进行filter，然后在查询的时候定位到精准的特征呢，或者是自动分配权重

**Comment by Mikezhang001** [comment]
写的很好，小白也能看明白一点。

**Comment by xuzhijvn** [comment]
感谢分享，过滤那一节没看懂，能结合一些举例就更好了

**Comment by HelloRabbits** [comment]
写的太好了

**Comment by sterio-wang** [comment]
Milvus怎么样？另外推荐用Redis + 插件还是单独的vdb比较好呢

**Comment by Fj-ivy** [comment]
深入浅出，分析到位！

**Comment by therainisme** [comment]
写得超级好！！！

![c01ec4e4fdbcacb81572c7e8970ad17d](https://github.com/user-attachments/assets/abfb6fa1-1cc2-4bb3-8e35-541689793fb1)

**Comment by tangkunyin** [comment]
老板辛苦了🍗🍗🍗

**Comment by ifengkou** [comment]
高质量文章，俩个b站视频也超高质量

**Comment by zuoniduimian** [comment]
学习了

**Comment by livingz** [comment]
大佬部分图是手绘的吗

**Comment by 0x3st** [comment]
以为是最近才有的文章，没想到是好久之前的了。只能说向量数据库的热度才刚刚开始。

之前看到视频说数据库的最终归宿都是 RDB，因为关系型数据库可以适配其他数据库的所有功能，甚至说没了 RDB 所有场景都是特殊场景。但我觉得对于专门的领域应该还得是专门的数据库来处理。不能因为地铁系统完善就砍掉了其他公交方式。

**Comment by p3han** [comment]
向量数据库
gpt缺点：输入端限制长度，token贵类似内存，想少一点token。长期记忆会丢失
解决方案：用向量数据库，减少无关上下文的输入， 向量数据库不是为了减少token数量本身，而是为了只喂最有用的token，让token“花得值”，提升效率、节省开销、提高准确率。  比如问清朝皇帝是谁，就不用把上下五千年全发过去。
原理：输入转换成向量vector embedding，先去做相似性搜索，有K-means,Faiss,product quantization,Hierachical Navigable Small Worlds,Locality Sensitive Hashing, 再计算向量之间的相似度，线性代数知识cos、欧几里得、点积。
Faiss：向量想象为包含在 Voronoi 单元格中 - 当引入一个新的查询向量时，首先测量其与质心 (centroids) 之间的距离，然后将搜索范围限制在该质心所在的单元格内。搜索几个质心是超参数。
product quantization：维度太高，每一簇的中心点指数增加，都不如原向量直接存，因为维度高导致的，就把每个高维向量分段，变成低维的，分段求中心点，每个中心店标号，就减少存储
HNSW：分层的图结构搜索算法，顶层的向量距离远，分大类型，底层很细致，适合精搜，上面先找，找到最近的，就选择这个，然后进入下一层。可以找到局部最优。
LSH：给空间随机划线，离得近的向量被分在直线同一边的概率大，设计专门的hash函数，目的就是为了碰撞，从而将相似的向量分到一起。来一个向量做搜索，先哈希，得到对应桶，在这个桶里搜索。空间维度太大，就把原始空间中的向量投影到一个目标维度的随机超平面，做起来就是用目标维度的随机投影矩阵与原始矩阵相乘。多次投影，将多次投影的结果组合成向量的低维表示，然后将这个低维表示用hash到一个桶中

**Comment by FallonXC** [comment]
膜拜大神

**Comment by jiangren1254** [comment]
这才是真正的好文章，必须点赞

**Comment by yyyanghj** [comment]
写的太好了！
